### PR TITLE
[otbn,crypto] Create .bss section for OTBN and use for RSA

### PR DESCRIPTION
**Pull request by jadephilipoom**
This PR allows OTBN programs to specify data in a `.bss` section instead of a `.data` section in order to indicate that the data does not have to be initialized; only the size and symbol locations will be included in the resulting binary, not the code in between. 

This PR also uses the new `.bss` section on the RSA-3072 code used by mask ROM sigverify (mask ROM is sensitive to code and those buffers are a lot of space to spend on zeroes; that was the original motivation for this change!) The size of the `.elf` file for sigverify's OTBN application (`run_rsa_verify_3072_rr_modexp`) went from 11036 to 6488 bytes as a result of this change. That's honestly a bit of a surprisingly large reduction to me (I'd expect it to save ~1.5k bytes, not 4.5k), but the tests are passing. If anyone could educate me on that one, I'd be curious to know! :slightly_smiling_face: 

I also checked locally that the OTBN standalone tests still work; there's no issue with `.weak` symbols in `.bss` being overridden by symbols in `.data` that do need to be initialized.

### CODE REVIEW

File: hw/ip/otbn/data/otbn.ld.tpl [hunk 1]
```diff
@@ -52,7 +52,16 @@ SECTIONS
 
         *(.data*)
         . = ALIGN(32);
+
+        /* Align section end (see note in .text section) */
+        . = ALIGN(4);
+
+    } >dmem AT>dmem_load
+
+    .bss : ALIGN(32)
```
Comment (alphan) [inline hw/ip/otbn/data/otbn.ld.tpl:61]: I think we should add a comment here explaining that this section is for input/output buffers and why `_dmem_end` is defined here instead of the `.data` section above.

I'm not very familiar with otbn build process but it's not immediately clear to me why `.bss` does not end up in the final binary since it is assigned to the same memory region as `.data` (`>dmem AT>dmem_load`). `otbn_build.py` explicitly removes `.scratchpad` but I couldn't find any references to `.bss`. Also, should this section be `NOLOAD`? @rswarbrick

**Comment by alphan** [comment]
Happy to see that this works! This is much cleaner than the `alloc`s in #9786.

> The size of the `.elf` file for sigverify's OTBN application (`run_rsa_verify_3072_rr_modexp`) went from 11036 to 6488 bytes as a result of this change. That's honestly a bit of a surprisingly large reduction to me (I'd expect it to save ~1.5k bytes, not 4.5k), but the tests are passing. If anyone could educate me on that one, I'd be curious to know! 🙂

I took a quick look and `run_rsa_verify_3072_rr_modexp` is built using `run_rsa_verify_3072_rr_modexp.s`, `rsa_verify_3072_rr.s`, and `rsa_verify_3072.s` and all of these files define buffers in the `.data` section. I didn't add up all the bytes but this may help explain the difference in size that you are seeing.

**Comment by jadephilipoom** [comment]
> I took a quick look and `run_rsa_verify_3072_rr_modexp` is built using `run_rsa_verify_3072_rr_modexp.s`, `rsa_verify_3072_rr.s`, and `rsa_verify_3072.s` and all of these files define buffers in the `.data` section. I didn't add up all the bytes but this may help explain the difference in size that you are seeing.

They define buffers, but all except the ones in `run_rsa_verify_3072_rr_modexp.s` are declared `.weak`, so I'd expect them to be overridden by the ones in that file and never defined -- if this didn't happen, I'd expect to get an error because it would exceed available non-scratchpad DMEM space!

> I'm not very familiar with otbn build process but it's not immediately clear to me why .bss does not end up in the final binary since it is assigned to the same memory region as .data (>dmem AT>dmem_load). otbn_build.py explicitly removes .scratchpad but I couldn't find any references to .bss. Also, should this section be NOLOAD?

My understanding here is that the `.bss` section does end up in the binary, in a way, but it's correctly marked as the BSS section of the ELF file and only the size of it is stored, not the actual data. I think it works without `NOLOAD` here becasue `.bss` is considered special and handled automatically; if I change `.bss` to `.foo` everywhere, it appears to get dropped from the binary entirely (presumably I'd have to do some further adjustments).

Here are some printouts from `riscv32-unknown-elf-size` -- they show exactly what I'd expect, 1600 bytes of data moving from the "data" to the "bss" segment of the ELF file. (Note that the "data" and "bss" in this printout are not derived from the definitions in `otbn.ld`, they're just the kinds of segments all ELFs have; if I make the `.data` section `NOLOAD`, it actually ends up under `bss` here.) Here's the printout from `master`:
```shell
jadep@jadep-l:~/opentitan$ /tools/riscv/bin/riscv32-unknown-elf-size -d ~/opentitan/build-out/sw/otbn/run_rsa_verify_3072_rr_modexp.elf 
   text	   data	    bss	    dec	    hex	filename
   1084	   1600	      0	   2684	    a7c	/home/jadep/opentitan/build-out/sw/otbn/run_rsa_verify_3072_rr_modexp.elf
```
...and here's this PR:
```shell
jadep@jadep-l:~/opentitan$ /tools/riscv/bin/riscv32-unknown-elf-size -d ~/opentitan/build-out/sw/otbn/run_rsa_verify_3072_rr_modexp.elf 
   text	   data	    bss	    dec	    hex	filename
   1084	      0	   1600	   2684	    a7c	/home/jadep/opentitan/build-out/sw/otbn/run_rsa_verify_3072_rr_modexp.elf
```

**Comment by jadephilipoom** [comment]
I realized upon closer inspection that I wasn't actually making the `.rv32embed.o` file smaller, because one of the arguments to `riscv32-unknown-elf-objcopy` in `otbn-build.py` sets all sections to `load` (and removes `scratchpad`, so it doesn't get overridden there). I've updated the PR so that these arguments now set only the `.data` section to `load`, to avoid overriding the default `noload` for BSS.

File: hw/ip/otbn/data/otbn.ld.tpl [hunk 2]
```diff
@@ -52,7 +52,16 @@ SECTIONS
 
         *(.data*)
         . = ALIGN(32);
+
+        /* Align section end (see note in .text section) */
+        . = ALIGN(4);
```
Comment (rswarbrick) [inline hw/ip/otbn/data/otbn.ld.tpl:57]: I don't think this is needed: the previous line aligns to 32 bytes, right?

File: hw/ip/otbn/data/otbn.ld.tpl (hunk 1 above)
Comment (rswarbrick) [reply hw/ip/otbn/data/otbn.ld.tpl:61]: I would have expected `.bss` to end up in the final binary, but it will still be a BSS section (as seen by the linker for the host binary): maybe we're managing to end up with memsz > filesz in the resulting host binary? This will be heavily dependent on where the OTBN image lands in the Ibex image.

To get a handle on what's going on, try linking the Ibex binary with `-Map`, which should show which input section is ending up where.

In particular, there's no magic handling for ".text" or ".data" in otbn_build.py. The only special section is `.scratchpad`, which gets chopped out.

**Review by rswarbrick — APPROVED** [review]
Looks reasonable but (like @alphan) I don't quite understand why this works! Would be good to get to the bottom of what's going on and write it down in a comment too.

File: hw/ip/otbn/data/otbn.ld.tpl (hunk 2 above)
Comment (jadephilipoom) [reply hw/ip/otbn/data/otbn.ld.tpl:57]: This was already in the `.data` section -- the diff is just showing it here because I added an identical clause to `.bss`. Perhaps it's not needed in either?

Comment (rswarbrick) [reply hw/ip/otbn/data/otbn.ld.tpl:57]: Yep, but the code used to be:
```
        *(.data*)
        . = ALIGN(32);
        *(.bss*)

        /* Align section end (see note in .text section) */
        . = ALIGN(4);
```
and your patch removes the `*(.bss*)` line which means there's no longer any way of "unaligning" things after the `ALIGN(32)`.

Comment (jadephilipoom) [reply hw/ip/otbn/data/otbn.ld.tpl:57]: Ah, I see -- I'll fix that, thanks!

**Comment by jadephilipoom** [comment]
> To get a handle on what's going on, try linking the Ibex binary with -Map, which should show which input section is ending up where.

Thanks for the tip to look at the map.  It looks like the build automatically produces some map files, such as `build-out/sw/device/silicon_creator/mask_rom/mask_rom_fpga_cw310.map`, so I was able to inspect those for the final mask ROM binary as well. Unfortunately, this PR reduces the size of the `.rv32embed.o` and `.elf` files produced from the OTBN code, but currently does not reduce the size of the final mask ROM binary.

The reason is that sections called `.bss*` are handled specially by `riscv-unknown-elf-ld`, which knows to put these in the BSS section of output ELF files. The `util/otbn_build.py` script prefixes the `.bss` section so it becomes `.rodata.otbn.bss`, which doesn't get recognized as a BSS section later in the mask ROM linker script (which matches everything with `.rodata*` and puts it in a loaded `.rodata` section):
https://github.com/lowRISC/opentitan/blob/a510f8e1f1ae7efd0f6924d376cea58295f2ce3b/sw/device/silicon_creator/mask_rom/mask_rom.ld#L108-L119

To solve this, we'll need to either change the prefix or change the rules in the mask ROM linker script. My suggestion is to change the renaming in `otbn_build.py` so that it continues to rename `.data.` and `.text` to `.rodata.otbn.data` and `.rodata.otbn.text`, but renames `.bss` to `.bss.otbn.bss` (since it's not really "read-only data" anway). This will match the existing `.bss*` rule in the mask ROM linker script and place it in the BSS section. WDYT?

**Comment by jadephilipoom** [comment]
A couple observations after looking at this some more:
- The loader at `sw/device/lib/runtime/otbn.c` isn't involved for mask ROM; it uses separate code at `sw/device/silicon_creator/lib/drivers/otbn.c`, which has a `otbn_zero_dmem` function; looks like the current code calls it when an app is initialized, then copies over the app's DMEM contents based on the `_dmem_end` symbol. So to avoid copying zeroes, we could change the `_dmem_end` symbol to `_data_end` or something; however, this is at runtime, so I'm not sure it affects binary size.
- I was looking at the size of the mask ROM elf file, `mask_rom_fpga_cw310.elf`, when I said the current version doesn't change the size; however, this code _does_ reduce the size of `mask_rom_fpga_cw310.bin` by exactly the expected 1600 bytes. @alphan , which do we care about? (Not certain why the `.elf` size isn't changing; `riscv32-unknown-elf-size` shows 1600 bytes moving from text to bss, as expected. But if we care about the `.bin`, it may not be worth debugging.)

**Comment by alphan** [comment]
> * One local change I've made doesn't reduce the size of the mask ROM elf file, `mask_rom_fpga_cw310.elf`; however, this code _does_ reduce the size of `mask_rom_fpga_cw310.bin` by exactly the expected 1600 bytes. @alphan , which do we care about? (Not certain why the `.elf` size isn't changing; `riscv32-unknown-elf-size` shows 1600 bytes moving from text to bss, as expected. But if we care about the `.bin`, it may not be worth debugging.)

Ultimately we care about the bin file, so it's good that its size decreases. You mentioned that 1600 bytes move from text to bss in the elf file. How does the file size and mem size of these change? @mundaym can you help with @jadephilipoom's question?

**Comment by jadephilipoom** [comment]
I've worked on this most of today and I think I'm on the path to a working solution, although I haven't finished it yet. Quick summary: Just renaming the sections isn't enough; the OTBN drivers (for both mask ROM and `runtime`) try to copy over all the data between the `_dmem_start` and `_dmem_end` symbols. Because the data and bss and sections will no longer be continuous when the mask rom linker script is applied, these drivers need to be updated to copy only between the start/end of the loaded data section, and also to allow DMEM writes to both the data and bss sections.

**Comment by jadephilipoom** [comment]
Should be working now, I've finished fixing up the loaders. `stat` tells me that `mask_rom_fpga_cw310.bin` is 1568 bytes smaller as a result of this change, which I'm guessing is the 1600 bytes that moves to BSS plus some overhead for having extra sections and symbols.

**Review by mcy — APPROVED** [review]
Ok. this looks correct to me. That was a doozy. Nice job Jade!

File: hw/ip/otbn/data/otbn.ld.tpl (hunk 1 above)
Comment (alphan) [reply hw/ip/otbn/data/otbn.ld.tpl:60]: `_dmem_bss_end` is clear enough, resolving this.

File: sw/device/silicon_creator/lib/otbn_util.c [hunk 3]
```diff
@@ -24,10 +24,25 @@ void otbn_init(otbn_t *ctx) {
...
+  if (app_dmem_data_start_addr <= ptr_addr &&
+      ptr_addr < app_dmem_data_end_addr) {
+    // Pointer is in the `data` section, which is at the start of DMEM
+    *dmem_addr_otbn = ptr_addr - app_dmem_data_start_addr;
+  } else if (app_dmem_bss_start_addr <= ptr_addr &&
+             ptr_addr < app_dmem_bss_end_addr) {
+    // Pointer is in the `bss` section, which is after the data section in DMEM
+    *dmem_addr_otbn = ptr_addr - app_dmem_bss_start_addr + app_dmem_bss_offset;
+  } else {
+    // Pointer is not in a valid DMEM region
```
Comment (alphan) [inline sw/device/silicon_creator/lib/otbn_util.c:43]: Not sure if I'm missing something but dmem_bss comes immediately after dmem_data, right? Can we use this to unify the checks, i.e. `app_dmem_data_start_addr <= ptr < app_dmem_bss_end_addr`, and simplify the pointer computation, i.e. `*dmem_addr_otbn = ptr_addr - app_dmem_data_start_addr`? Then we wouldn't need `dmem_bss_offset` here.

File: util/otbn_build.py [hunk 4]
```diff
@@ -201,7 +201,8 @@ def main() -> int:
         out_embedded_obj = out_dir / (app_name + '.rv32embed.o')
         args = (['-O', 'elf32-littleriscv',
                  '--prefix-sections=.rodata.otbn',
-                 '--set-section-flags=*=alloc,load,readonly',
+                 '--set-section-flags=*=alloc,readonly',
+                 '--set-section-flags=.data=load',
```
Comment (rswarbrick) [inline util/otbn_build.py:204]: I don't think I understand this. I would have thought you want every section to be ALLOC, LOAD, READONLY except for the new BSS section, which should be ALLOC, READONLY.

If I've understood right, this sets `.text` to ALLOC, READONLY (what happened to LOAD?!) and `.data` to LOAD (it should also be READONLY).

Doesn't something like this work?
```
                 '--set-section-flags=*=alloc,load,readonly',
                 '--set-section-flags=.bss=alloc,readonly',
```

File: sw/device/lib/runtime/otbn.c [hunk 5]
```diff
@@ -17,14 +17,29 @@ const int kOtbnWlenBytes = 256 / 8;
...
+    // Pointer is in the data section, which is at the start of DMEM
+    *dmem_addr_otbn = ptr_addr - app_dmem_data_start_addr;
+  } else if (app_dmem_bss_start_addr <= ptr_addr &&
+             ptr_addr < app_dmem_bss_end_addr) {
+    // Pointer is in the bss section, which is after the data section
+    *dmem_addr_otbn = ptr_addr - app_dmem_bss_start_addr + app_dmem_bss_offset;
+  } else {
+    // Pointer is not in a valid DMEM region
+    return kOtbnBadArg;
+  }
```
Comment (rswarbrick) [inline sw/device/lib/runtime/otbn.c:41]: This could be dramatically simplified if you could just assume that OTBN .bss came just after OTBN .data in the Ibex link. Is that not true?

File: sw/device/lib/runtime/otbn.h [hunk 6]
```diff
@@ -32,13 +32,30 @@ typedef struct otbn_app {
...
+   * End of uninitialized OTBN data memory.
    */
-  const uint8_t *dmem_start;
+  const uint8_t *dmem_bss_end;
   /**
-   * End of OTBN data memory.
+   * Offset of the BSS section within DMEM.
    */
-  const uint8_t *dmem_end;
+  const uint8_t *dmem_bss_offset;
```
Comment (rswarbrick) [inline sw/device/lib/runtime/otbn.h:58]: This should be a uint32_t, I think: it's an absolute value giving an OTBN address.

File: sw/device/lib/runtime/otbn.c [hunk 7]
```diff
@@ -63,15 +78,20 @@ otbn_result_t otbn_init(otbn_t *ctx, mmio_region_t base_addr) {
 }
 
 otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t app) {
-  if (app.imem_end <= app.imem_start || app.dmem_end < app.dmem_start) {
+  if (app.imem_end <= app.imem_start ||
+      app.dmem_data_end < app.dmem_data_start ||
+      app.dmem_bss_end < app.dmem_bss_start ||
+      (uintptr_t)app.dmem_bss_offset <
+          (uintptr_t)app.dmem_data_end - (uintptr_t)app.dmem_data_start) {
```
Comment (rswarbrick) [inline sw/device/lib/runtime/otbn.c:85]: I think these checks are:
- The IMEM range isn't backwards
- The DMEM data range isn't backwards
- The DMEM BSS range isn't backwards
- The BSS offset lies above the top of initialised DMEM.

which looks good to me.  I wonder whether it makes sense to define a "`check_wellformed_range`" helper function to do the first three checks. It would be kind of trivial and get inlined, but it would help to document what's going on. Alternatively, you could split this into 4 `if` statements with comments. But that's a bit sad because you have to repeat the return statement.

Also, if you change `dmem_bss_offset` to a `uint32_t` as I suggested in another comment then I think you can get rid of all the casts here: the pointers being subtracted are `uint8_t*`, so subtraction will give the byte difference.

The same comments apply on the `silicon_creator` side.

File: sw/device/lib/runtime/otbn.c [hunk 8]
```diff
@@ -83,9 +103,16 @@ otbn_result_t otbn_load_app(otbn_t *ctx, const otbn_app_t app) {
...
   }
 
-  if (dmem_size > 0) {
-    if (dif_otbn_dmem_write(&ctx->dif, 0, ctx->app.dmem_start, dmem_size) !=
-        kDifOk) {
+  // Zero all of DMEM
+  otbn_result_t err = otbn_zero_data_memory(ctx);
+  if (err != kOtbnOk) {
+    return err;
+  }
```
Comment (rswarbrick) [inline sw/device/lib/runtime/otbn.c:110]: Oh! If you're doing this then you don't actually have to track the BSS section at all, right? I think this is a totally reasonable approach but, in this case, please can we ditch `dmem_bss_*` from the context object.

File: sw/device/silicon_creator/lib/otbn_util.c (hunk 3 above)
Comment (rswarbrick) [reply sw/device/silicon_creator/lib/otbn_util.c:43]: Hah! I just wrote the same thing on the runtime code :-D

File: util/otbn_build.py [hunk 9]
```diff
@@ -200,9 +200,12 @@ def main() -> int:
         sym_pfx = '_otbn_app_{}_'.format(app_name)
         out_embedded_obj = out_dir / (app_name + '.rv32embed.o')
         args = (['-O', 'elf32-littleriscv',
-                 '--prefix-sections=.rodata.otbn',
                  '--set-section-flags=*=alloc,readonly',
                  '--set-section-flags=.data=load',
+                 '--rename-section=.text=.rodata.otbn.text',
+                 '--rename-section=.start=.rodata.otbn.start',
+                 '--rename-section=.data=.rodata.otbn.data',
+                 '--rename-section=.bss=.bss.otbn.bss',
```
Comment (rswarbrick) [inline util/otbn_build.py:208]: Hmm, why is this needed?

**Review by rswarbrick — CHANGES_REQUESTED** [review]

File: sw/device/silicon_creator/lib/otbn_util.c (hunk 3 above)
Comment (jadephilipoom) [reply sw/device/silicon_creator/lib/otbn_util.c:43]: The bss section comes directly after the data section in DMEM address space, but not in Ibex address space (because they end up in different memory sections as defined by the mask ROM linker script, i.e. the `.data.` and `.bss` section there, and there are other files' data and bss portions in between them).

For example, if the OTBN program has a 4-byte buffer called `foo` that's stored in `data` and a 512-byte buffer called `bar` that's stored in `bss`, we have something like the following in Ibex address space:
```
0x100 .data
.... <other programs' data sections> ...
0x200 _otbn_run_rsa_verify_3072_rr_modexp_dmem_data_start (app_dmem_data_start_addr)
0x200 _otbn_run_rsa_verify_3072_rr_modexp_foo
0x204 _otbn_run_rsa_verify_3072_rr_modexp_dmem_data_end (app_dmem_data_end_addr)
.... <other programs' data sections> ...
0x400 .bss
.... <other programs' bss sections> ...
0x500 _otbn_run_rsa_verify_3072_rr_modexp_dmem_bss_start (app_dmem_bss_start_addr)
0x500 _otbn_run_rsa_verify_3072_rr_modexp_bar
0x1012 _otbn_run_rsa_verify_3072_rr_modexp_dmem_bss_end (app_dmem_bss_end_addr)
```
In DMEM address space, we want to have something like:
```
0x00 .data
0x00 foo
0x20 .bss
0x20 baz
```
Note that the `.bss` section gets aligned to 32 bits there, even though we only have 4 bytes of `.data`. This depends on the OTBN linker script (which forces both `.data` and `.bss` to be 32-byte aligned) and _also_ on `.balign` directives in the OTBN program; if `baz` had a `.balign 64`, for instance, the whole bss section would start at 0x40 in DMEM address space. And from the context of the mask ROM linker script, it's impossible to know that from symbol addresses alone!

That's why the OTBN linker script records `dmem_bss_offset`; in order to know the offset _in DMEM address space_, we need to know how much space there is between `_dmem_data_start` and `_dmem_bss_start` when they're right next to each other (i.e. in the context of `otbn.ld`, not `mask_rom.ld`).

File: util/otbn_build.py (hunk 4 above)
Comment (jadephilipoom) [reply util/otbn_build.py:204]: Yes, it does! I misunderstood the behavior of `set-section-flags`, and thought it would set everything to `alloc, readonly` and set `.data` additionally to `load` (but it's true `.text` should be `load` also). Your suggestion works locally; I'll update the PR to match.

File: sw/device/silicon_creator/lib/otbn_util.c (hunk 3 above)
Comment (rswarbrick) [reply sw/device/silicon_creator/lib/otbn_util.c:43]: Oh, now I understand. Makes sense, thanks.

File: sw/device/lib/runtime/otbn.h (hunk 6 above)
Comment (rswarbrick) [reply sw/device/lib/runtime/otbn.h:58]: Oh good point. But can't you just do the cast to a `uint32_t` in the body of OTBN_APP_T_INIT?

File: sw/device/lib/runtime/otbn.c (hunk 8 above)
Comment (rswarbrick) [reply sw/device/lib/runtime/otbn.c:110]: I think that just means you need the BSS offset, right?

File: util/otbn_build.py (hunk 9 above)
Comment (rswarbrick) [reply util/otbn_build.py:208]: Ah, I see. Presumably the problem is that if the OTBN BSS ends up mixed in with all the other rodata then it takes up address space (even though it's NOLOAD). Probably the resulting ELF has a gap between the segments, but that doesn't help because it has to be zero-padded when generating the Ibex bin file. (Is that right?)

Have you tried making the .bss section not allocatable? I think this might be an easier approach, that also won't require Ibex to zero out a big wodge of its own memory at runtime.

File: sw/device/lib/runtime/otbn.c (hunk 8 above)
Comment (jadephilipoom) [reply sw/device/lib/runtime/otbn.c:110]: I need the BSS offset to do the actual conversion, but `otbn_ptr_to_dmem_addr` should also return an "invalid argument" error if the Ibex pointer passed as input isn't in one of the two Ibex memory sections (data/bss) that point to DMEM for the given app. So to check that the Ibex pointer is valid, I need the bss start/end addresses.

File: util/otbn_build.py (hunk 9 above)
Comment (jadephilipoom) [reply util/otbn_build.py:208]: > Ah, I see. Presumably the problem is that if the OTBN BSS ends up mixed in with all the other rodata then it takes up address space (even though it's NOLOAD). Probably the resulting ELF has a gap between the segments, but that doesn't help because it has to be zero-padded when generating the Ibex bin file. (Is that right?)

Yes, that's exactly right. Everything that ends up in the rodata section ends up getting loaded, including the OTBN BSS if it's there, and the binary size doesn't decrease.

> Have you tried making the .bss section not allocatable? I think this might be an easier approach, that also won't require Ibex to zero out a big wodge of its own memory at runtime.

I tried setting the flags so `.bss` is not allocatable, but it still seems to be allocated in the resulting binary based on the size. For clarity, I changed the arguments here to:
```
                 '--set-section-flags=*=alloc,load,readonly',
                 '--set-section-flags=.bss=noload,readonly',
                 '--rename-section=.text=.rodata.otbn.text',
                 '--rename-section=.start=.rodata.otbn.start',
                 '--rename-section=.data=.rodata.otbn.data',
                 '--rename-section=.bss=.rodata.otbn.bss',
                 '--remove-section=.scratchpad',
```
Is that what you meant?

File: sw/device/lib/runtime/otbn.c (hunk 8 above)
Comment (rswarbrick) [reply sw/device/lib/runtime/otbn.c:110]: Understood, but I'm not sure that feature really matters all that much. I'd probably just suggest you check against the size of the DMEM. That said, if the NOALLOC flag I suggested fixes things, then everything will be contiguous again...

File: sw/device/lib/runtime/otbn.h (hunk 6 above)
Comment (jadephilipoom) [reply sw/device/lib/runtime/otbn.h:58]: Good point, that works! I do have to remove the `const` qualifier from this line for it to compile, though, otherwise the `ctx->app = app;` line from `otbn_load_app` fails with `cannot assign to non-static data member 'app' with const-qualified data member 'dmem_bss_offset'`. It seems to me like that change is probably worth it; just noting it here in case there are objections/new ideas for how to get around that.

File: sw/device/silicon_creator/lib/otbn_util.c (hunk 3 above)
Comment (alphan) [reply sw/device/silicon_creator/lib/otbn_util.c:43]: Thanks for the clarification @jadephilipoom!

File: sw/device/lib/runtime/otbn.c (hunk 8 above)
Comment (jadephilipoom) [reply sw/device/lib/runtime/otbn.c:135]: I was mistaken in my earlier comment -- I do also need `dmem_bss_start` to compute the DMEM address (because the computation for a bss symbol is basically `ibex_ptr - dmem_bss_start + bss_offset`. I could eliminate `dmem_bss_end` and assume that anything past `dmem_bss_start` and not in `data` range is in bss (because we don't necessarily have the guarantee that `data` will come first if the mask ROM linker script changes!). However, I'm not sure that it would be an improvement; it seems a bit harder to understand than having the same logic for both data and bss, and would only save one field from the context object.

Unfortunately the noalloc suggestion doesn't work, assuming that I understood it correctly -- `set-section-flags` doesn't have an option for `noalloc`, but I tried removing `alloc` from `.bss`, adding `noload`, and putting `NOALLOC` above it in `otbn.ld`; none of these work to reduce the size of the binary. My understanding is that it just gets overridden by the mask ROM linker script, which says `.rodata` is always loaded.

Comment (rswarbrick) [reply sw/device/lib/runtime/otbn.c:135]: Ah, good point about `dmem_bss_start`. At this point I agree: not worth trying to remove!

I'm still slightly confused about the noalloc thing: I rather expected it to work! I'm going to try fiddling around locally a bit and see whether I can get things right.

**Comment by rswarbrick** [comment]
Ok, I think I've understood what's going on and my "can't you just do..." comments ain't gonna work :-) So I think we should go with this approach for now: I'll have a last pass through the PR and then approve.

**BUT**: It's still a bit silly. In particular, we're allocating a big slab of zeros in the Ibex bss section that we'll never actually use. That doesn't cost us anything in the binary, but it does cost data memory on the host CPU. I've done a bit of thinking, and I think I've got a plan for how to do this more nicely. Basically, we run objcopy twice (with some judicious renamings). The first run generates something like what we've got already, but drops both .bss and .scratchpad. The second run uses the `--extract-symbol` argument to generate an ELF with just symbols, giving OTBN-side addresses. Once we link everything together, we can copy the data and text segments across as before. But we drop the `otbn_ptr_to_dmem_addr` function entirely, doing DMEM address translation statically by just using the two symbols. I'll have a play locally and see if I can get everything working: I think it should be a bit simpler than Philipp's (admittedly clever!) approach that we've been using up to now.

File: hw/ip/otbn/data/otbn.ld.tpl (hunk 2 above)
Comment (rswarbrick) [reply hw/ip/otbn/data/otbn.ld.tpl:55]: The original reason was to allow people to use e.g. `BN.LID` on the last stuff in the section without having to worry about getting integrity errors from uninitialised memory just above. But (a) Don't do that! and (b) We're probably zeroing all of memory in the first place anyway, so it won't be a problem.

Anyway, that's a long-winded way of saying "sounds good to me" :-)

File: util/otbn_build.py (hunk 9 above)
Comment (rswarbrick) [reply util/otbn_build.py:208]: I think that `--prefix-sections=.rodata.otbn` might be easier, but let's worry about that in a follow-up PR :-)

File: sw/device/lib/runtime/otbn.h (hunk 6 above)
Comment (rswarbrick) [reply sw/device/lib/runtime/otbn.h]: Yep, a `const uint32_t` is a 32-bit integer that you can't assign to. A `const uint8_t*` is a slangy shorthand for `uint8_t const*` (if I've got that right!) which means a pointer you can't write **through**.

**Review by rswarbrick — APPROVED** [review]

**Comment by rswarbrick** [comment]
Right, I've put my master plan into action! See #10105.
